OpenSim/Simulation/Model/PathPoint.h

```diff
@@ -86,68 +65,106 @@ OpenSim_DECLARE_CONCRETE_OBJECT(PathPoint, Object);
+    void setBody(const PhysicalFrame& body) {
+        this->setParentFrame(body);
+    }
+
+    void changeBodyPreserveLocation(const SimTK::State& s, PhysicalFrame& body){
+        if (!hasParent()) {
+            throw Exception("PathPoint::changeBodyPreserveLocation attempted to "
+                " change the body on PathPoint which was not assigned to a body.");
+        }
+        // if it is already assigned to aBody, do nothing
+        const PhysicalFrame& currentFrame = getParentFrame();
+
+        if (currentFrame == body)
+            return;
+
+        // Preserve location means to switch bodies without changing
+        // the location of the point in the inertial reference frame.
+        upd_location() = currentFrame.findLocationInAnotherFrame(s, get_location(), body);
+
+        // now assign this point's body to point to aBody
+        setParentFrame(body);
+    }
 
-    const PhysicalFrame& getBody() const { return *_body; }
-    const std::string& getBodyName() const { return _bodyName; }
-    GeometryPath* getPath() const { return _path; }
+    const PhysicalFrame& getBody() const { return getParentFrame(); }
+    const std::string& getBodyName() const { return getParentFrame().getName(); }
+
+    GeometryPath* getPath() const { return _path.get(); }
+
+    virtual void scale(const SimTK::State& s, const SimTK::Vec3& scaleFactors) {
+        for (int i = 0; i < 3; i++)
+            upd_location()[i] *= scaleFactors[i];
+    }
 
-    virtual void scale(const SimTK::State& s, const SimTK::Vec3& aScaleFactors);
     virtual const WrapObject* getWrapObject() const { return NULL; }
 
     virtual bool isActive(const SimTK::State& s) const { return true; }
-    virtual void connectToModelAndPath(const Model& aModel, GeometryPath& aPath);
+    virtual void connectToModelAndPath(Model& aModel, GeometryPath& aPath);
     virtual void update(const SimTK::State& s) { }
 
     // get the relative velocity of the path point with respect to the body
     // it is connected to.
-    virtual void getVelocity(const SimTK::State& s, SimTK::Vec3& aVelocity);
+    virtual void getVelocity(const SimTK::State& s, SimTK::Vec3& velocity) {
+        velocity[0] = velocity[1] = velocity[2] = 0.0;
+    }
     // get the partial of the point location w.r.t. to the coordinates (Q)
     // it is dependent on.
     virtual SimTK::Vec3 getdPointdQ(const SimTK::State& s) const
         { return SimTK::Vec3(0); }
 
-    virtual void updateGeometry();
+    virtual void updateGeometry() {}
 
     // Utility
-    static PathPoint* makePathPointOfType(PathPoint* aPoint, const std::string& aNewTypeName);
-    static void deletePathPoint(PathPoint* aPoint) { if (aPoint) delete aPoint; }
+    static PathPoint* makePathPointOfType(PathPoint* aPoint,
+        const std::string& aNewTypeName);
 
-protected:
+    static void deletePathPoint(PathPoint* aPoint) { if (aPoint) delete aPoint; }
```

Comment:
Yes @chrisdembia this function is ised by the muscle/path editor and is exercised in the tutorial2 script as well.